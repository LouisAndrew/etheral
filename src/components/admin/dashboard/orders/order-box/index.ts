import { OrderBox } from './order-box';

export default OrderBox;
