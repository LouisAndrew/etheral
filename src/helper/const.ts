import { Product } from './schema/product';
import { Order } from './schema/order';

const mockProduct1: Product = {
    pid: '1234',
    name: 'Mock product',
    slug: 'ctg1/product1',
    category: 'ctg1',
    idrPrice: 23000,
    collection: 'fall',
    productImages: [],
    urls: [],
    weight: 20,
    amount: 12,
};

export const mockProducts: Product[] = [mockProduct1];

export const mockOrder: Order = {
    oid: 'asdadwda',
    buyerName: 'Jane Doe',
    buyerEmail: 'jane@doe',
    buyerPhone: 123412412,
    buyerUId: 'admin',
    buyerAddr: 'Storkowerstrasse',
    buyerPostal: 1234143,
    total: 32000,
    currency: 'IDR',
    date: new Date(),
    via: 'va',
    channel: 'bni',
    products: [{ pid: mockProduct1.pid, amount: 12 }],
    paid: false,
    delivered: false,
    transactionData: {
        sessionId: '1234',
        paymentNo: '12',
        paymentName: 'Jane Doe - BNI',
        expired: '1231241412',
        fee: 3000,
    },
    shippingMethod: 'OKE',
};

export const mockOrders: Order[] = [mockOrder];

export const mockOrderShipped: Order = {
    ...mockOrder,
    delivered: true,
    shippingData: {
        shippedDate: new Date(),
        trackingCode: '12342',
        shippedBy: 'admin1',
    },
};
