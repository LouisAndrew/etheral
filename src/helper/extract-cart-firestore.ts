import { intersectionBy } from 'lodash';

import { SetCartArgs } from 'state/actions/cart';
import { InCart } from './schema/firebase-user';
import { Product } from './schema/product';

export default ({
    firestoreCartData,
    allProducts,
}: {
    firestoreCartData: InCart;
    allProducts: Product[];
}): SetCartArgs => {
    const inCartProduct = intersectionBy(allProducts, firestoreCartData, 'pid');
    return inCartProduct.map(product => ({
        product,
        amount: intersectionBy(firestoreCartData, [product], 'pid')[0].amount, // get the intersection between the product -> pid is unique, so the function is guaranteed to return one item.
    }));
};
