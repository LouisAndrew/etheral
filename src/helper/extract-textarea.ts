import { snakeCase } from 'lodash';

/**
 * function to extract options from a textarea
 * @param textarea textarea input (e.g sizes: S, M, L)
 */
export const extractTextArea = (textarea: string) =>
    textarea.split(',').map(text => snakeCase(text).toUpperCase());
