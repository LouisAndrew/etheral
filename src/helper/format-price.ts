export const formatPrice = (price: number | string) => {
    const value =
        typeof price === 'string'
            ? parseFloat(price.replace('IDR ', ''))
            : price;
    return new Intl.NumberFormat('id').format(value);
};
