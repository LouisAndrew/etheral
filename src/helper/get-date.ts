const pad2 = (n: number): string => (n < 10 ? `0${n}` : n.toString());

const getDate = (date: Date) => {
    const year = date.getUTCFullYear();
    const month = pad2(date.getUTCMonth() + 1); // getMonth() is zero-indexed, so we'll increment to get the correct month number
    const day = pad2(date.getUTCDate());
    const hours = pad2(date.getUTCHours() + 7);
    const minutes = pad2(date.getUTCMinutes());

    return year + '-' + month + '-' + day + ' ' + hours + ':' + minutes;
};

const getDateIpaymu = (date: Date) => {
    const year = date.getUTCFullYear();
    const month = pad2(date.getUTCMonth() + 1); // getMonth() is zero-indexed, so we'll increment to get the correct month number
    const day = pad2(date.getUTCDate());
    const hours = pad2(date.getUTCHours() + 7); // UTC offset for WIB => +7
    const minutes = pad2(date.getUTCMinutes());
    const seconds = pad2(date.getUTCSeconds());

    return `${year}${month}${day}${hours}${minutes}${seconds}`;
};

const monthNames = [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December',
];

const getDateReadable = (date: Date) => {
    const year = date.getUTCFullYear();
    const month = date.getUTCMonth();
    const day = date.getUTCDate();

    return `${day} ${monthNames[month]} ${year}`;
};

export { getDate, getDateIpaymu, getDateReadable };
