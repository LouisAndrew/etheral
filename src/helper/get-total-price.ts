import { Product } from 'helper/schema';
import { withDiscount } from './with-discount';
import { Currencies } from 'state/reducers/currency-reducer';

type HelperArgs = {
    [key: string]: any;
    amount: number;
    product: Product;
};

const getTotalPriceIdr = (args: HelperArgs[]) => {
    return args.reduce((acc, curr) => {
        const { amount, product } = curr;

        // destructure prices prop
        const {
            prices: { discountPercentage, idrPrice },
        } = product;

        // get price.
        const price =
            discountPercentage > 0
                ? withDiscount(idrPrice, discountPercentage)
                : idrPrice;

        return acc + amount * price;
    }, 0);
};

const getTotalPriceAud = (args: HelperArgs[]) => {
    return args.reduce((acc, curr) => {
        const { amount, product } = curr;

        // destructure prices prop
        const {
            prices: { discountPercentage, ausPrice },
        } = product;

        // get price.
        const price =
            discountPercentage > 0
                ? withDiscount(ausPrice, discountPercentage)
                : ausPrice;

        return acc + amount * price;
    }, 0);
};

/**
 * function to get price (formatted ?) based on currency
 * @param args product name and prices + its discount percentage.
 * @param currency current currency state
 */
const getPrice = (args: HelperArgs, currency: Currencies) => {
    const {
        product: { prices },
        amount,
    } = args;

    const { discountPercentage, idrPrice, ausPrice } = prices;

    // handle IDR.
    if (currency === Currencies.IDR) {
        const price =
            discountPercentage > 0
                ? withDiscount(idrPrice, discountPercentage)
                : idrPrice;

        return amount * price;
    } else {
        const price =
            discountPercentage > 0
                ? withDiscount(ausPrice, discountPercentage)
                : ausPrice;

        return amount * price;
    }
};

export { getTotalPriceIdr, getTotalPriceAud, getPrice };
