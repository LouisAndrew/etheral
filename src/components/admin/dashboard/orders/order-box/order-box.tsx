import React from 'react';

import { Box, Text, Flex } from 'rebass';

import { Currencies } from 'state/reducers/currency-reducer';
import { getDate } from 'helper/get-date';
import StatusBadge, { BadgeTypes } from '../status-badge';

type Props = {
    oid: string;
    date: Date;
    paid: boolean;
    shipped: boolean;
    currency: Currencies;
    bg: string;
};

const OrderBox: React.FC<Props> = ({
    oid,
    date,
    paid,
    shipped,
    currency,
    bg,
}) => {
    const textStyling = {
        fontFamily: 'body',
        fontSize: [0, 0, 1],
        color: '#222',
        fontWeight: 'bold',
        css: `
			 white-space: nowrap;
			overflow: hidden;
			text-overflow: ellipsis;
		`,
    };

    return (
        <Box
            bg={bg}
            sx={{
                display: 'grid',
                gridTemplateColumns: 'repeat(4, minmax(64px, 1fr))',
                gridGap: 2,
                transition: '0.2s',
                '&:hover': {
                    cursor: 'pointer',
                },
            }}
            px={[2]}
            py={[2]}
        >
            <Text sx={{ gridColumn: '1/2' }} {...textStyling}>
                {oid}
            </Text>
            <Flex flexWrap="wrap" sx={{ gridColumn: '2/3' }}>
                <StatusBadge type={BadgeTypes.PAYMENT} paid={paid} />
                <StatusBadge type={BadgeTypes.SHIPPING} shipped={shipped} />
            </Flex>
            <Text sx={{ gridColumn: '3/4' }} {...textStyling}>
                {currency}
            </Text>
            <Text sx={{ gridColumn: '4/5' }} {...textStyling}>
                {getDate(date)}
            </Text>
        </Box>
    );
};

export { OrderBox };
