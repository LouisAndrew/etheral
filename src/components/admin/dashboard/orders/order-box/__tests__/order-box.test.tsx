import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';

import { mockOrder } from 'helper/const';
import { OrderBox } from '../order-box';

describe('OrderBox', () => {
    const { oid, date, delivered: shipped, currency, paid } = mockOrder;

    const mockFocusOrder = jest.fn(() => {});
    const mockBg = '#fafafa';

    const Element = (
        <OrderBox
            oid={oid}
            date={date}
            shipped={shipped}
            currency={currency}
            focusOrder={mockFocusOrder}
            bg={mockBg}
            paid={paid}
        />
    );

    afterEach(cleanup);

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(Element, div);
    });

    it('should react to click event correctly', () => {
        const { getByTestId } = render(Element);
        const OrderBox = getByTestId('order-box');

        userEvent.click(OrderBox);
        expect(mockFocusOrder).toBeCalled();
    });

    it('should render with background color correctly', () => {
        const { getByTestId } = render(Element);
        const OrderBox = getByTestId('order-box');

        expect(OrderBox).toHaveStyle(`background-color: ${mockBg}`);
    });

    it('matches snapshot', () => {
        const run = true;

        if (run) {
            const tree = renderer.create(Element).toJSON();
            expect(tree).toMatchSnapshot();
        }
    });
});
