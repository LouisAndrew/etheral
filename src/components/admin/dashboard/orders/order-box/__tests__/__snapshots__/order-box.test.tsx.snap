// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderBox matches snapshot 1`] = `
<div
  className="css-1drooja"
  data-testid="order-box"
  onClick={[Function]}
>
  <div
    className="css-te8eeg"
  >
    asdadwda
  </div>
  <div
    className="css-vtr4jd"
  >
    <div
      className="css-1a0lovt"
    >
      Not Paid
    </div>
    <div
      className="css-1a0lovt"
    >
      Not Shipped
    </div>
  </div>
  <div
    className="css-fjt3vn"
  >
    IDR
  </div>
  <div
    className="css-1wi5gb3"
  >
    2020-10-05 18:57
  </div>
</div>
`;
